module/zfs/dsl_bookmark.c

```diff
@@ -413,6 +427,148 @@ dsl_bookmark_create(nvlist_t *bmarks, nvlist_t *errors)
 	    fnvlist_num_pairs(bmarks), ZFS_SPACE_CHECK_NORMAL));
 }
 
+typedef struct dsl_bookmark_clone_arg {
+	nvlist_t *dbcc_bmarks;
+	nvlist_t *dbcc_errors;
+} dsl_bookmark_clone_arg_t;
+
+static int
+dsl_bookmark_clone_check_impl(nvpair_t *pair, dmu_tx_t *tx)
+{
+	dsl_pool_t *dp = dmu_tx_pool(tx);
+	int error;
+	zfs_bookmark_phys_t bmark_phys = { 0 };
+
+	/* Verify that the new bookmark does not already exist */
```

Comment:
Good catch, no, the clone code didn't.
I'll add some tests for this, too.

BTW: the `create` code does the check, but only very implicitly because `dsl_bookmark_create_check_impl` does the bookmark-existence check based on the shortname, relative to the snapshot's dataset:

https://github.com/zfsonlinux/zfs/blob/03b516cf864230375dd5077c4792716322895010/module/zfs/dsl_bookmark.c#L140-L160

Is such implicit checking considered good enough for the DSL, or should I add the checks I'll add to the bookmark clone path to the bookmark create path as well?

Reply:
( Code block for this is `/* Verify that new and target bookmark are on the same dataset */` above )

Reply:
I think that implicit checking is fine, for the create path. Might be good to add a comment that calls out that it does that, in case some future refactor accidentally breaks it.

Reply:
The relevant check in dsl_bookmark_create_check_impl() is dsl_dataset_is_before(), which I would say is relatively explicit.  It's comment says, `Return TRUE if 'earlier' is an earlier snapshot in 'later's timeline.`